http/peer-id-auth.md

```diff
@@ -0,0 +1,101 @@
+   * The signature is over the concatenated result of:
+   ```
+     <varint-length> + "origin=" + server-name + 
+     <varint-length> + "client-challenge=" + base64-encoded-client-chosen-client-challenge + 
+     <varint-length> + "challenge=" + base64-encoded-challenge
+   ```
+   * The client chosen client-challenge MUST be randomly generated.
+   * The client chosen client-challenge SHOULD be at least 32 bytes.
+   * The client MUST use the same server-name as what is used for the TLS
+     session.
+1. The server MUST verify the signature using the server name used in the TLS
+   session. The server MUST return 401 Unauthorized if the server fails to
+   validate the signature.
+1. If the signature is valid, the server has authenticated the client's peer id
+   and MAY fulfill the request according to application logic. If the request is
+   fulfilled, the server sets the `Authentication-Info` response header to the
+   following:
+    ```
+    Libp2p-Challenge peer-id="<encoded-peer-id-bytes>",sig="<base64-signature-bytes>"
+    ```
+   * The signature is over the concatenated result of:
+        ```
+        <varint-length> + "origin=" + server-name + 
+        <varint-length> + "client-challenge=" + base64-encoded-client-chosen-client-challenge + 
+        <varint-length> + "client=" + <encoded-client-peer-id-bytes>
+        ```
+1. The client can then authenticate the server with the the signature from
+   `Authentication-info`.
+
+## Server Authentication
+
+Clients may wish to only authenticate the server's peer ID, but not themselves.
+For example, a short lived client may want to get a block from a specific peer.
+
+The protocol to do so is as follows:
+
+1. The client should set the request header `Authorization` to
+   `Libp2p-Challenge-Server-Only <base64-encoded-client-chosen-client-challenge>`.
+1. The server should response to the request and set `Authentication-Info`
+   response header to the following:
+    ```
+    Libp2p-Challenge-Server-Only peer-id="<encoded-peer-id-bytes>",sig="<base64-signature-bytes>"
+    ```
+   * The signature is over the concatenated result of:
+        ```
+        <varint-length> + "origin=" + server-name + 
+        <varint-length> + "client-challenge=" + base64-encoded-client-chosen-client-challenge
+        ```
+1. The client can now authenticate the server.
+
+## Authentication Endpoint
+
+Because the client needs to make a request to authenticate the server, and the
+client may not want to make the real request before authenticating the server,
+the server MAY provide an authentication endpoint. This authentication endpoint
+is like any other application protocol, and it shows up in `.well-known/libp2p`,
+but it only does the authentication flow. It doesn’t send any other data besides
+what is defined in the above authentication flows. The protocol id for the
+authentication endpoint is `/http-peer-id-auth/1.0.0`.
```

Comment:
Should we specify that the client SHOULD NOT send any payload with the HTTP request?